01_stratified_classification/nbconverted/lasso_range_gene_optimizers.py

```diff
@@ -0,0 +1,791 @@
+    g.axes[1].set_xlim((10e-7, 10))
+    g.set_ylabels(f'{metric.upper()}')
+    sns.move_legend(g, "center", bbox_to_anchor=[1.035, 0.55], frameon=True)
+    g._legend.set_title('Dataset')
+    new_labels = ['train', 'holdout', 'test']
+    for t, l in zip(g._legend.texts, new_labels):
+        t.set_text(l)
+    g.set_titles('Optimizer: {col_name}')
+     
+    plt.suptitle(f'LASSO parameter vs. {metric.upper()}, {plot_gene}', y=1.0)
+
+plt.tight_layout()
+
+if output_plots:
+    plt.savefig(os.path.join(output_plots_dir, f'{gene}_parameter_vs_perf.svg'), bbox_inches='tight')
+
+
+# In[9]:
+
+
+ll_nz_coefs_df['optimizer'] = 'liblinear'
+sgd_nz_coefs_df['optimizer'] = 'SGD'
+
+nz_coefs_df = pd.concat((ll_nz_coefs_df, sgd_nz_coefs_df))
+nz_coefs_df.head()
+
+
+# In[10]:
+
+
+perf_coefs_df = (plot_df
+    .merge(nz_coefs_df,
+           left_on=['gene', 'optimizer', 'lasso_param', 'seed', 'fold'],
+           right_on=['gene', 'optimizer', 'lasso_param', 'seed', 'fold'])
+)
+
+print(perf_coefs_df.shape)
+perf_coefs_df.head()
+
+
+# In[11]:
+
+
+sns.set({'figure.figsize': (10, 5)})
+sns.set_style('ticks')
+sns.histplot(perf_coefs_df.nz_coefs)
+plt.title(f'Nonzero coefficient distribution for {plot_gene}')
+plt.xlabel('Number of nonzero coefficients')
+
+linear_bins_df = []
+quantiles_df = []
+
+ax = plt.gca()
+for q in np.linspace(0.1, 0.9, 9):
+    quantiles_df.append([q, perf_coefs_df.nz_coefs.quantile(q)])
+    ax.axvline(x=perf_coefs_df.nz_coefs.quantile(q),
+                      color='black', linestyle='--')
+    
+for b in np.linspace(0, perf_coefs_df.nz_coefs.max(), 11):
+    ax.axvline(x=b, color='grey', linestyle=':')
+    
+# create custom legend for bin boundary lines
+legend_handles = [
+    Line2D([0], [0], color='black', linestyle='--'),
+    Line2D([0], [0], color='grey', linestyle=':'),
+]
+legend_labels = ['deciles', 'linear bins']
+l = ax.legend(legend_handles, legend_labels, title='Bin type',
+              loc='lower left', bbox_to_anchor=(1.01, 0.4))
+ax.add_artist(l)
+    
+quantiles_df = pd.DataFrame(quantiles_df, columns=['quantile', 'value'])
+quantiles_df
+
+
+# In[12]:
+
+
+# TODO: figure out 0 quantile general solution
+perf_coefs_df['nz_linear_bin'] = pd.cut(
+    perf_coefs_df.nz_coefs,
+    bins=np.linspace(0, perf_coefs_df.nz_coefs.max(), 11),
+    labels=[f'{q}' for q in range(1, 11)],
+    include_lowest=True
+)
+
+print(perf_coefs_df.nz_linear_bin.unique().sort_values())
+perf_coefs_df.head()
+
+
+# In[13]:
+
+
+ll_param_bin_map = (perf_coefs_df
+    [perf_coefs_df.optimizer == 'liblinear']
+).loc[:, ['lasso_param', 'nz_linear_bin']]
+
+(ll_param_bin_map
+   .drop_duplicates()
+   .groupby('lasso_param')
+)['nz_linear_bin'].apply(list)
+
+
+# In[14]:
+
+
+sgd_param_bin_map = (perf_coefs_df
+    [perf_coefs_df.optimizer == 'SGD']
+).loc[:, ['lasso_param', 'nz_linear_bin']]
+
+(sgd_param_bin_map
+   .drop_duplicates()
+   .groupby('lasso_param')
+)['nz_linear_bin'].apply(list)
+
+
+# In[15]:
```

Comment:
This whole section needs some explanation. What are you trying to see with these plots?